FRC: Retrieval Checking Requirements (#1089)
filecoin-project/FIPs: FRCs/frc-retrieval-checking-requirements.md

    1. Spark typically repeats every retrieval check 40-100 times. Scanning CAR byte range 40-100 times does not bring enough value to justify the network bandwidth & CPU cost.

 1. It's not clear how can retrieval checkers discover the address where the SP serves piece retrievals.

## Backwards Compatibility
<!--All FIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The FIP must explain how the author proposes to deal with these incompatibilities. FIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->

[Retrieval Requirements](#retrieval-requirements) document the current status and remove Graphsync and Bitswap protocols. Existing miner operations need to enable/configure [IPFS Trustless HTTP Gateway protocol](https://specs.ipfs.tech/http-gateways/trustless-gateway/) retrievals to meet the new requirements.

|Miner Software|Supports HTTP retrievals|Notes
|-|:-:|-|
|Boost|✅| Manual setup required: [docs](https://boost.filecoin.io/retrieving-data-from-filecoin/http-retrieval#payload-retrievals-car-and-raw).

> |Curio|✅| TODO: OOTB or manual setup?

Author: Thank you for a quick response.

I am not sure if we are talking about the same thing here.

Let's say I have a cat video `bafy...` that I store in a Filecoin deal with an SP running Curio. Then I want to retrieve back my video by making an HTTP request to `https://{storage-provider}/cid/bafy...`, using the [IPFS Trustless HTTP Gateway](https://specs.ipfs.tech/http-gateways/trustless-gateway/) protocol.

IIUC, you are saying that:
- The SP must configure Curio to accept HTTP connections, otherwise it cannot accept incoming deals
- By default, Curio provides payload retrievals via the Trustless HTTP GW protocol.
- As a result, in the default setup, the SP will serve me back my cat video at `https://{storage-provider}/cid/bafy...`

Could you please confirm?

> |Curio|✅| TODO: OOTB or manual setup?

Reviewer: Yes. That is correct. Without HTTP config, Curio cannot make deals, publish IPNI etc. So this is not an optional requirement. Curio only supports HTTP retrievals.

> |Venus Droplet| ? | TODO: OOTB or manual setup?

Author: I tried to find the answer in Droplet's code, it seems it advertises Graphsync retrievals only.

https://github.com/ipfs-force-community/droplet/blob/f4bffcde1b689119688fa1e64c19387ea0f51d73/indexprovider/index_provider.go#L418-L430

> |Venus Droplet| ? | TODO: OOTB or manual setup?

Author: Droplet docs on configuring HTTP retrievals:

- https://github.com/ipfs-force-community/droplet/blob/master/docs/zh/模拟官方机器人HTTP方式检索.md#开启-http-检索

- Translation to EN: https://github-com.translate.goog/ipfs-force-community/droplet/blob/master/docs/zh/模拟官方机器人HTTP方式检索.md?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp#开启-http-检索

> |Venus Droplet| ? | TODO: OOTB or manual setup?

Author: Add to the spec document in 1b8f592

This FRC enables retrieval checks based on payload sampling with support for all kinds of storage deals (f05, direct data onboarding, etc.).

The service-level indicators produced by retrieval checker networks can be integrated into incentive mechanisms like FIL+ or paid storage deals to drive improvements in the availability and reliability of retrieval service offered by individual SPs and Filecoin as a whole.

## Implementation
<!--The implementations must be completed before any core FIP is given status "Final", but it need not be completed before the FIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->

### Storage Provider Software

|Requirement|Boost|Curio|Venus
|-|:-:|:-:|:-:
|Advertises payload retrieval to IPNI|✅|✅|✅
|Trustless HTTP GW retrievals|✅|✅|?|
|Link on-chain MinerId and IPNI provider identity|✅|❌|✅
|Construct IPNI ContextID from (PieceCID, PieceSize)|❌|✅|❌

### IPNI Reverse Index

> Status: design phase

Author: @masih do you have any tracking GH issue for your work on the IPNI reverse index? I'd like to add link to that issue here.

Reviewer 2: https://github.com/ipni/roadmap/issues/1

    "ProviderResults": [{
      "ContextID": "ZnJpc2JpaQ==",
      "Metadata": "oBIA",
      "Provider": {
        "ID": "12D3KooWC8gXxg9LoJ9h3hy3jzBkEAxamyHEQJKtRmAuBuvoMzpr",
        "Addrs": [
          "/dns/frisbii.fly.dev/tcp/443/https"
        ]
      }
    }]
  }]
}

```

### Retrieval Requirements

1. Whenever a deal is activated, the SP MUST advertise all IPFS/IPLD payload block CIDs found in the Piece to IPNI. See the [IPNI Specification](https://github.com/ipni/specs/blob/main/IPNI.md) and [IPNI HTTP Provider](https://github.com/ipni/specs/blob/main/IPNI_HTTP_PROVIDER.md) for technical details.

2. Whenever SP stops storing a Piece (e.g. because the last deal for the Piece has expired or was slashed), the SP SHOULD advertise removal of all payload block CIDs included in this Piece.

3. The SP MUST provide retrieval of the IPFS/IPLD payload blocks via the [IPFS Trustless HTTP Gateway protocol](https://specs.ipfs.tech/http-gateways/trustless-gateway/).

### Retrieval Checking Requirements

In addition to the above [retrieval requirements](#retrieval-requirements), SPs are asked to meet the following:

#### Link on-chain MinerId and IPNI provider identity

> Storage providers are requires to use the same libp2p peer ID for their block-chain identity as returned by `Filecoin.StateMinerInfo` and for the index provider identity used when communicating with IPNI instances like [cid.contact](https://cid.contact).

Reviewer: This requirement cannot be fulfilled in Curio. We no longer have a concept of minerID <> Unique peerID binding. IPNI must be extended to support other keys types like worker key to sign ads.

Author: I am aware of that; see the note in the text below this paragraphs.

```markdown
> [!NOTE]
> This is open to extensions in the future, we can support more than one form of linking
> index-provides to filecoin-miners. See e.g. [ipni/spec#33](https://github.com/ipni/specs/issues/33).
```

From my point of view, I prefer not to block progress on this FRC until the Curio team figures out how to extend IPNI to support other key types. Instead, I'd like this FRC to document the solution that works with Boost & Venus now and then enhance it with the new mechanism Curio needs once that new solution is agreed on.

Reviewer 3: I am probably mistaken here but Droplet (Venus' Boost) supports multiple minerIDs being associated with a single PeerID ([see docs](https://droplet.venus-fil.io/operation/#droplet-proxy-libp2p-listener)), does that mean if I am using Droplet, I need to limit myself to a 1:1 relationship to meet this requirement?

Author: Great call, @lanzafame! I am still learning more about how Venus Droplet work and what features they offer.

Based on the docs you linked to, I believe you can have multiple minerIDs associated with a single Droplet PeerID and still meet this requirement.

In Spark, we need the PeerID returned by `Filecoin.StateMinerInfo` to match the PeerID used in IPNI advertisements. Spark does not check whether that PeerID is unique or shared by multiple miners.

Author: Referencing for visibility - we are discussing a possible solution for this problem here:

https://github.com/filecoin-project/curio/pull/377

- Because such ContextIDs are scoped per piece (not per deal), miner software must de-duplicate advertisements for deals storing the same piece.

## Test Cases

<!--Test cases for an implementation are mandatory for FIPs that are affecting consensus changes. Other FIPs can choose to include links to test cases if applicable.-->

Not applicable, but see the examples in [Specification](#specification).

## Security Considerations
<!--All FIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. FIP submissions missing the "Security Considerations" section will be rejected. A FIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.-->

_TODO: add more details._

> We trust SPs to honestly advertise Piece payload blocks to IPNI. Attack vector: a malicious SP can always advertise the same payload block for all pieces persisted.

Author: > We trust SPs to honestly advertise Piece payload blocks to IPNI. Attack vector: a malicious SP can always advertise the same payload block for all pieces persisted.

@willscott do you remember how we wanted to mitigate this attack vector?

Author: I described building a small network of validator nodes to audit compliance.

---
fip: "<to be assigned>"
title: Retrieval Checking Requirements
author: "Miroslav Bajtoš (@bajtos)"
discussions-to: https://github.com/filecoin-project/FIPs/discussions/1086
status: Draft
type: FRC
created: 2024-12-02
# spec-sections:
#   - <section-id>
#   - <section-id>
# requires (*optional): <FIP number(s)>
# replaces (*optional): <FIP number(s)>
---

<!--You can leave these HTML comments in your merged FIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new FIPs. Note that a FIP number will be assigned by an editor. When opening a pull request to submit your FIP, please use an abbreviated title in the filename, `fip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

# FIP-Number: Retrieval Checking Requirements

## Simple Summary

<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the FIP.-->

> To make Filecoin a usable data storage offering, we need the content to be retrievable. It's difficult to improve what you don't measure; therefore, we need to measure the quality of retrieval service provided by each storage provider. To allow 3rd-party networks like [Spark](https://filspark.com) to sample active deals from the on-chain activity and check whether the SP is serving retrievals for the stored content, we need SPs to meet the following requirements:

Reviewer 4: My quibble with this paragraph is that you're asserting a _need_ for "3rd-party networks", yet the structure you outline below isn't the only way to achieve the goal you've stated (because it's a "need"). This is more like Spark's preferred path, which is fine, it's just not a generalisable _need_. Maybe it'd be best to not give the impression to the reader that this is the only way to do this.

I'd prefer a more direct approach here: _To allow Spark to sample deals, we need the following to be true._

It's nice to have a standard that's specified via FRC but it's not like we have a queue of retrieval checkers waiting for such a standard in order to get going.

> To make Filecoin a usable data storage offering, we need the content to be retrievable. It's difficult to improve what you don't measure; therefore, we need to measure the quality of retrieval service provided by each storage provider. To allow 3rd-party networks like [Spark](https://filspark.com) to sample active deals from the on-chain activity and check whether the SP is serving retrievals for the stored content, we need SPs to meet the following requirements:
> 
> 1. Link on-chain MinerId and IPNI provider identity ([spec](#link-on-chain-minerid-and-ipni-provider-identity)).
> 2. Provide retrieval service using the [IPFS Trustless HTTP Gateway protocol](https://specs.ipfs.tech/http-gateways/trustless-gateway/).
> 3. Advertise retrievals to IPNI.
> 4. In IPNI advertisements, construct the `ContextID` field from `(PieceCID, PieceSize)` ([spec](#construct-ipni-contextid-from-piececid-piecesize))
> 
> Meeting these requirements needs support in software implementations like Boost, Curio & Venus Droplet but potentially also updates in settings configured by the individual SPs.

Reviewer 5: ```suggestion
To make Filecoin a usable data storage offering, we need the content to be retrievable. It's difficult to improve what you don't measure; therefore, we need to measure the quality of retrieval service provided by each storage provider. This FRC outlines requirements that SPs and their software stacks should meet to allow 3rd-party networks to sample active deals from the on-chain activity and check whether the SP is serving retrievals for the stored content.
```

The goal here is not to go into technical detail. I left a non-binding suggestion; something along these lines would be preferable.

The content in 26-29 would potentially be a good fit for the abstract; see comment below.

1. A client wanting to download the dataset identified by CID `bafybei(...)` queries an IPNI instance like [cid.contact](https://cid.contact) to find the nodes providing retrievals service for this dataset.

2. The client picks a retrieval provider that supports the [IPFS Trustless HTTP Gateway protocol](https://specs.ipfs.tech/http-gateways/trustless-gateway/).

3. The client requests the content for CID `bafybei(...)` at the URL (multiaddr) specified by the [IPNI provider result](https://github.com/ipni/specs/blob/12482e4e1bd92a7c6c079bf23f2533a4ddb9e363/IPNI.md#json-find-response) of the selected provider.

Example IPNI `ProviderResult` describing retrieval provider offering IPFS Trustless HTTP Gateway retrievals:

```json
{
  "MultihashResults": [{
    "Multihash": "EiAT38UKZPlJfhyZQH8cAMNjUPeKBfQn6HMdiqGZ2xJicA==",
    "ProviderResults": [{
      "ContextID": "ZnJpc2JpaQ==",
      "Metadata": "oBIA",
      "Provider": {
        "ID": "12D3KooWC8gXxg9LoJ9h3hy3jzBkEAxamyHEQJKtRmAuBuvoMzpr",
        "Addrs": [
          "/dns/frisbii.fly.dev/tcp/443/https"
        ]
      }
    }]
  }]
}
```

### Retrieval Requirements

> 1. Whenever a deal is activated, the SP MUST advertise all IPFS/IPLD payload block CIDs found in the Piece to IPNI. See the [IPNI Specification](https://github.com/ipni/specs/blob/main/IPNI.md) and [IPNI HTTP Provider](https://github.com/ipni/specs/blob/main/IPNI_HTTP_PROVIDER.md) for technical details.

Reviewer 3: Can a client opt out of their deal payload being indexing?

Reviewer 4: It'll depend on the market software, Boost should support optionality here ([see](https://github.com/filecoin-project/boost/pull/1051)) and I imagine Curio will follow suit because there's already been demonstrated demand for "private" deals.

Reviewer 6: this is all in the context of data that's claimed to be 'retrievable'

We're not defining a standard here for signalling between client and SP / deal making for how a client communicates to an SP that data should be retrievable or not.

```json5
{
  "MultihashResults": [
    {
      "Multihash": "EiAT38UKZPlJfhyZQH8cAMNjUPeKBfQn6HMdiqGZ2xJicA==",
      "ProviderResults": [
        // (...)
        {
          "ContextID": "AXESIFVcxmAvWdc3BbQUKlYcp2Z2DuO2w5Fo4jmIC8IbMX00",
          "Metadata": "oBIA",
          "Provider": {
            "ID": "12D3KooWPNbkEgjdBNeaCGpsgCrPRETe4uBZf1ShFXStobdN18ys",
            "Addrs": [
              "/dns/cesginc.com/tcp/443/https"
            ]
          }
        }
      ]
    }
  ]
}
```

#### Construct IPNI `ContextID` from `(PieceCID, PieceSize)`

The advertisements for IPNI must deterministically construct the `ContextID` field from the public deal metadata - the tuple `(PieceCID, PieceSize)` - as follows:

- Use DAG-CBOR encoding ([DAG-CBOR spec](https://ipld.io/specs/codecs/dag-cbor/spec/))
- The piece information is serialised as an array with two items:
   1. The first item is the piece size represented as `uint64`
   2. The second item is the piece CID represented as a custom tag `42`
- In places where the ContextID is represented as a string, convert the CBOR bytes to string using the hex encoding.

>    _Note: the Go module https://github.com/ipni/go-libipni handles this conversion automatically._

Author: FIXME: IPNI uses base64 encoding (not hex encoding).

Author: ```suggestion
- In places where the ContextID is represented as a string, convert the CBOR bytes to string using the base64 encoding.
   _Note: the Go module https://github.com/ipni/go-libipni handles this conversion automatically._
```

A reference implementation of this serialization algorithm in Go is maintained in [https://github.com/filecoin-project/go-state-types/](https://github.com/filecoin-project/go-state-types/blob/32f613e4d4450b09da3c81982dd6d7dba9c6f6f2/abi/cbor_gen.go#L23-L48).

**Example**

Input:

```json5
{
  "PieceCID": "baga6ea4seaqpyzrxp423g6akmu3i2dnd7ymgf37z7m3nwhkbntt3stbocbroqdq",
  "PieceSize": 34359738368 // 32 GiB
}
```

> Output - ContextID (hex-encoded, split into two lines for readability):
> 
> ```
> 821B0000000800000000D82A5828000181E203922020FC66377F35
> B3780A65368D0DA3FE1862EFF9FB36DB1D416CE7B94C2E1062E80E
> ```

Author: ````suggestion
Output - ContextID (base64-encoded):

```
ghsAAAAIAAAAANgqWCgAAYHiA5IgIPxmN381s3gKZTaNDaP+GGLv+fs22x1BbOe5TC4QYugO
```
````

scan in every retrieval checker node is redundant and wasteful. It also makes it more involved to
implement alternate retrieval checker networks.

IPNI already contains a list of payload CIDs contained in every advertised Filecoin deal, therefore
we propose leveraging this existing infrastructure.

- Storage Providers keep advertising payload CIDs to IPNI as they do now.
- IPNI implements reverse index lookup allowing retrieval checkers to sample CIDs in a given deal.
- Retrieval checkers can easily get a sample of payload CIDs and then retrieve only the selected CID(s).
- Retrieval checking does not add any extra network bandwidth overhead to SPs beyond the actual retrieval.

### Alternatives considered

Assuming there is a limit on the size of any CAR block in a Piece, it's possible to sample one payload block using the following algorithm:

1. Let's assume the maximum CAR block size is 4 MB and we have deal's `PieceCID` and `PieceSize`.
2. Pick a random offset $o$ in the piece so that $0 <= o <= PieceSize - 2*4 MB$.
3. Send an HTTP range-retrieval request to retrieve the bytes in the range`(o, o+2*4MB)`.

> 4. Parse the bytes to find a sequence that looks like a CARv1 block header.

Reviewer 4: Sequence that looks like a CARv1 will either start at offset `0` or be discoverable from a CARv2 header which is 51 bytes long from offset `0` or will alternatively be a PoDSI container which should also be discoverable from offset `0`.

5. Extract the CID from the CARv1 block header.

> 6. Hash the block's payload bytes and verify that the digest equals to the CID.

Reviewer 4: CID in a CARv1 block header isn't a CID of the payload, it's _typically_ (but not always, and may even be omitted) the root of the entire DAG within the CAR.

But the approach here could be generalised as something like the following: download a small chunk from the start of a piece, determine whether it's readable as either a CARv2, CARv1 or PoDSI container, then use that knowledge to read further sections of the containing CAR(s) to find CIDs to retrieve and build a catalog by progressive byte-range scanning of CAR section headers.

It's true that this is inefficient and would require a stored state to build up knowledge of a piece, but it is possible to do an IPLD block discovery by doing many, progressive, and small, piece retrievals if the SP is exposing the `/piece/` endpoint.

You're still back at the problem of trusting the SP that the "piece" they are serving you corresponds to the PieceCID which you trust. As long as the SP is responsible for serving you the response _or_ reporting CIDs to IPNI, you're at the mercy of the SP to tell you what a piece contains rather than the client who ought to have a canonical mapping of piece->blocks (or at least did have at the beginning). Or, to be properly trustless, downloading the entire piece and verifying the PieceCID matches the piece they gave you, but I guess you have to rule this option out and therefore be trusting the SP to play nice.

Reviewer 6: we should extend range reads over pieces to include the intermediate proof tree (which may already be in the podsci index, or can be generated on the side like the cid index) to allow validation that the range is indeed the expected sub-section of the overall PieceCID

The reasons why we rejected this approach:

 1. It's inefficient.

    1. Each retrieval check requires two requests - one to download ~8MB chunk of a piece, the second one to download the payload block found in that chunk.

    1. Spark typically repeats every retrieval check 40-100 times. Scanning CAR byte range 40-100 times does not bring enough value to justify the network bandwidth & CPU cost.

>  1. It's not clear how can retrieval checkers discover the address where the SP serves piece retrievals.

Reviewer 4: Just like this FRC is dictating a convention, you could dictate a convention that at least one of the on-chain `Multiaddrs` stored in the miner actor of the SP is where you could retrieve from.

(1)

We trust SPs to honestly advertise Piece payload blocks to IPNI. Attack vector: a malicious SP can always advertise the same payload block for all pieces persisted.

We see this as a broader problem with the lack of trust & verifiability in the IPNI architecture.
How can the network verify that parties storing content (Filecoin SPs, IPFS nodes) are honestly advertising
all content?

To flag Filecoin SPs that are not advertising all deal payload, we can build a small network of validator nodes performing the following algorithm:

1. Let's assume the maximum CAR block size is 4 MB and we have deal's `PieceCID` and `PieceSize`.

   _(E.g. validators can walk IPNI advertisement chains and extract piece information by parsing `ContextID`.)_

2. Pick a random offset $o$ in the piece so that $0 <= o <= PieceSize - (2*4 MB)$.

3. Send an HTTP range-retrieval request to retrieve the bytes in the range $[o, o+(2*4MB)]$.

   _We need the server to return an inclusion proof up to the PieceCID root._

> 4. Parse the bytes to find a sequence that looks like a CARv1 block header.

Reviewer 4: see my point above, this doesn't need to be a random offset
